augur/augur_client.py

```diff
@@ -311,5 +311,16 @@ def _contracts_from_abi(self, abi):
       filter['fromBlock'] = current_block
       filter['toBlock'] = current_block + increment
       events_filter = self._ethereum_client.eth.filter(filter)
-      events_filter.format_entry = handler
-      events_filter.get_all_entries()
+      events = events_filter.get_all_entries()
+      for event in events:
+        await handler(event)
+    try:
+      ethereum_uri = self._ethereum_client.providers[0].endpoint_uri
+      websocket = await websockets.connect(ethereum_uri)
+      filter['fromBlock'] = 'latest'
+      del filter['toBlock']
+      await self._send_request('eth_subscribe', websocket, ['logs', filter])
+      async for event in websocket:
```

Comment:
Why is this for loop ```async``` but the one on line 315 is not ```async```?

Reply:
The for loop on 315 is a list of events that the Web3py method `get_all_entries()` returns when it gets old events. The async on 323 waits on a response from the websocket. It uses `__aiter__` to iterate asynchronously through all messages that the `eth_subscribe` method outputs when a new event is found. Web3py has no method for 'subscribing' to Ethereum events, so I performed a JSON-RPC call to the node here.